BUG: F_INT type conflict with f2py translation of INTEGER type on i386

### Describe your issue.

Compiling scipy with GCC 14 on i386 fails.


### Reproducing Code Example

Just building scipy with a GCC 14 snapshot seems enough. I tried scipy-1.11.3-4.fc40 from Fedora rawhide.

The issue seems to be this in `scipy/interpolate/src/fitpack.pyf`:

```
#ifdef HAVE_ILP64
typedef npy_int64 F_INT;
#else
typedef npy_int32 F_INT;
#endif
```

This defines `F_INT` as `long int`, but f2py appears to translate

```
       integer intent(hide) :: iopt = 0
```

as plain `int` in `.mesonpy-…/scipy/interpolate/dfitpackmodule.c`:

```
    int iopt = 0;
```


### Error message


First error:

```
  ../scipy/interpolate/src/_fitpackmodule.c: In function ‘fitpack_bispev’:
  ../scipy/interpolate/src/_fitpackmodule.c:255:10: error: assignment to ‘npy_int32 *’ {aka ‘long int *’} from incompatible pointer type ‘int *’ [-Wincompatible-pointer-types]
    255 |     iwrk = (int *)(wrk + lwrk);
        |          ^
```

This one is easily patched:

```diff
diff --git a/scipy/interpolate/src/_fitpackmodule.c b/scipy/interpolate/src/_fitpackmodule.c
index 466879b8291249e6..e86bf08ba6ae4f01 100644
--- a/scipy/interpolate/src/_fitpackmodule.c
+++ b/scipy/interpolate/src/_fitpackmodule.c
@@ -252,7 +252,7 @@ fitpack_bispev(PyObject *dummy, PyObject *args)
         goto fail;
     }
     wrk = wa;
-    iwrk = (int *)(wrk + lwrk);
+    iwrk = (F_INT *)(wrk + lwrk);
     if (nux || nuy) {
         PARDER(tx, &nx, ty, &ny, c, &kx, &ky, &nux, &nuy, x, &mx, y, &my, z,
                 wrk, &lwrk, iwrk, &kwrk, &ier);
```

But then:

```
  scipy/interpolate/dfitpackmodule.c: In function ‘f2py_rout_dfitpack_fpcurf0’:
  scipy/interpolate/dfitpackmodule.c:2718:30: error: passing argument 1 of ‘f2py_func’ from incompatible pointer type [-Wincompatible-pointer-types]
   2718 |                 (*f2py_func)(&iopt,x,y,w,&m,&xb,&xe,&k,&s,&nest,&tol,&maxit,&k1,&k2,&n,t,c,&fp,fpint,wrk,wrk+nest,wrk+nest*k2,wrk+nest*2*k2,wrk+nest*3*k2,nrdata,&ier) ;
        |                              ^~~~~
        |                              |
        |                              int *
  scipy/interpolate/dfitpackmodule.c:2718:30: note: expected ‘F_INT *’ {aka ‘long int *’} but argument is of type ‘int *’
  scipy/interpolate/dfitpackmodule.c:2718:42: error: passing argument 5 of ‘f2py_func’ from incompatible pointer type [-Wincompatible-pointer-types]
   2718 |                 (*f2py_func)(&iopt,x,y,w,&m,&xb,&xe,&k,&s,&nest,&tol,&maxit,&k1,&k2,&n,t,c,&fp,fpint,wrk,wrk+nest,wrk+nest*k2,wrk+nest*2*k2,wrk+nest*3*k2,nrdata,&ier) ;
        |                                          ^~
        |                                          |
        |                                          int *
  scipy/interpolate/dfitpackmodule.c:2718:42: note: expected ‘F_INT *’ {aka ‘long int *’} but argument is of type ‘int *’
  scipy/interpolate/dfitpackmodule.c:2718:53: error: passing argument 8 of ‘f2py_func’ from incompatible pointer type [-Wincompatible-pointer-types]
   2718 |                 (*f2py_func)(&iopt,x,y,w,&m,&xb,&xe,&k,&s,&nest,&tol,&maxit,&k1,&k2,&n,t,c,&fp,fpint,wrk,wrk+nest,wrk+nest*k2,wrk+nest*2*k2,wrk+nest*3*k2,nrdata,&ier) ;
        |                                                     ^~
        |                                                     |
        |                                                     int *
  scipy/interpolate/dfitpackmodule.c:2718:53: note: expected ‘F_INT *’ {aka ‘long int *’} but argument is of type ‘int *’
  scipy/interpolate/dfitpackmodule.c:2718:59: error: passing argument 10 of ‘f2py_func’ from incompatible pointer type [-Wincompatible-pointer-types]
   2718 |                 (*f2py_func)(&iopt,x,y,w,&m,&xb,&xe,&k,&s,&nest,&tol,&maxit,&k1,&k2,&n,t,c,&fp,fpint,wrk,wrk+nest,wrk+nest*k2,wrk+nest*2*k2,wrk+nest*3*k2,nrdata,&ier) ;
        |                                                           ^~~~~
        |                                                           |
        |                                                           int *
  scipy/interpolate/dfitpackmodule.c:2718:59: note: expected ‘F_INT *’ {aka ‘long int *’} but argument is of type ‘int *’
  scipy/interpolate/dfitpackmodule.c:2718:70: error: passing argument 12 of ‘f2py_func’ from incompatible pointer type [-Wincompatible-pointer-types]
   2718 |                 (*f2py_func)(&iopt,x,y,w,&m,&xb,&xe,&k,&s,&nest,&tol,&maxit,&k1,&k2,&n,t,c,&fp,fpint,wrk,wrk+nest,wrk+nest*k2,wrk+nest*2*k2,wrk+nest*3*k2,nrdata,&ier) ;
        |                                                                      ^~~~~~
        |                                                                      |
        |                                                                      int *
```


### SciPy/NumPy/Python version and system information

Fedora 40/rawhide i686-linux-gnu with GCC 14.0.1 20240127 and scipy 1.11.3.

64-bit builds do not have this issue.

--- a/scipy/interpolate/src/_fitpackmodule.c
+++ b/scipy/interpolate/src/_fitpackmodule.c
@@ -28,8 +28,8 @@ static PyObject *fitpack_error;
 
 #else
 
-#define F_INT npy_int32
-#define F_INT_NPY NPY_INT32
+#define F_INT int
+#define F_INT_NPY NPY_INT
 #define F_INT_MAX NPY_MAX_INT32
 #if NPY_BITSOF_SHORT == 32
 #define F_INT_PYFMT   "h"

--- a/scipy/interpolate/src/fitpack.pyf
+++ b/scipy/interpolate/src/fitpack.pyf
@@ -16,7 +16,7 @@ python module dfitpack ! in
 #ifdef HAVE_ILP64
 typedef npy_int64 F_INT;
 #else
-typedef npy_int32 F_INT;
+typedef int F_INT;
 #endif
 
 static double dmax(double* seq, F_INT len) {

